src/metabase/search/scoring.clj

```diff
@@ -169,6 +172,7 @@
 (defn- combined-score
   [{:keys [text-score result]}]
   [(pinned-score result)
+   (dashboard-count-score result)
    (- text-score)
    (model->sort-position (:model result))
    (:name result)])
```

Comment:
I'm a bit confused why scores are being made negative. I'm assuming dashboard_count is positive and should boost scores. Making it negative will reduce the score right?

```clojure
((juxt first last)
 (transduce (map (fn [x] [[(- x) x] x]))
            accumulate-top-results
            (range 100)))
[[[-49 49] 49] [[0 0] 0]]

((juxt first last)
 (transduce (map (fn [x] [[x x] x]))
            accumulate-top-results
            (range 100)))
[[[50 50] 50] [[99 99] 99]]
```

Treating the score vector as a tuple of number of dashboards its in and then text score, it keeps the ones that are in the fewest dashboards right? We'd want [50 _] through [99 _] but the negative scores gives us [49 _] through [0 _] occurrences.

Also, if i'm understanding, this vector compare might give some unintended consequences. If something is in 5 dashboards and has a text score of 1 it will be highlighted over something that is in 0 dashboards but has a text score of 500. Might be a bit confusing why an exact text match is 5th in a list of search results when something that shares a single term and is in a bunch of dashboards is highlighted.

If i'm totally misunderstanding sorry for the noise but that's what i'm reading right now.

Reply:
No, you bring up a subtle bug. Good catch! To summarize:

1) The priority saves the _largest_ n items
2) Since `sort` puts small things first, I've been making a lot of things negative. Scores of 1, 3, and 6 would become -1, -3 and -6, so that when sorted they'd be in the order `[-6, -3, -1]`.
3) This breaks the logic of the PQ: if it only held two items, they'd be `[-3, -1]`, not `[-6, -3]`.

I just pushed up a commit that fixes it. Specifically, high scores now correspond to important with no negative magic, but there are a few `reverse`s put in to change the results from ascending order to descending order.

Also, this revealed a related bug I overlooked: order is not guaranteed when turning a PQ into another type of sequence, so it needs to be sorted anyway. So I pulled out this logic into the `top-results` function, and testing that instead of the accumulator led to stronger tests anyway.


-----

> Also, if i'm understanding, this vector compare might give some unintended consequences.

Kyle and I were aware of this, and it's not entirely certain that this is a bad thing. Our plan (which you had no way of knowing) is to ship the vector-based scoring, play with it on stats, and then decide if we should keep it as-is or blend together the numbers a bit more (so that, like you said, something that was a much better text match takes precedence over something in more dashboards. Defining how much is "much better" is obviously…not cut and dried.)